csharp_dotnetcore/Using-Cards/Using_Cards/CardsBot.cs

```diff
@@ -0,0 +1,339 @@
+
+        public CardsBot(CardsBotAccessors accessors)
+        {
+            _accessors = accessors;
+
+            Dialogs = new DialogSet(accessors.ConversationDialogState);
+            Dialogs.Add(new WaterfallDialog("cardSelector", new WaterfallStep[] { ChoiceCardStep, ShowCardStep }));
+            Dialogs.Add(new ChoicePrompt("cardPrompt"));
+        }
+
+        public async Task OnTurnAsync(ITurnContext turnContext, CancellationToken cancellationToken = new CancellationToken())
+        {
+            switch (turnContext.Activity.Type)
+            {
+                case ActivityTypes.Message:
+                    try
+                    {
+                        var dc = await Dialogs.CreateContextAsync(turnContext, cancellationToken);
+                        await dc.ContinueAsync(cancellationToken);
+
+                        if (!dc.Context.Responded)
+                        {
+                            await dc.BeginAsync("cardSelector", cancellationToken: cancellationToken);
+                        }
+                    }
+                    catch (Exception e)
+                    {
+                        Console.WriteLine(e);
+                        throw;
+                    }
+
+                    break;
+                case ActivityTypes.ConversationUpdate:
+
+                    // Send a welcome message to the user and tell them what actions they need to perform to use this bot
+                    if (turnContext.Activity.MembersAdded.Any())
+                    {
+                        {
+                            foreach (var member in turnContext.Activity.MembersAdded)
+                            {
+                                var newUserName = member.Name;
+                                if (member.Id != turnContext.Activity.Recipient.Id)
+                                {
+                                    await turnContext.SendActivityAsync($"Welcome to CardBot {newUserName}. This bot will show you different types of Rich Cards.  Please type anything to get started.", cancellationToken: cancellationToken);
+                                }
+                            }
+                        }
+                    }
+                    break;
+                default:
+                    // There is no code in this bot to deal with ActivityTypes other than conversationUpdate or message
+                    await turnContext.SendActivityAsync("This type of activity is not handled in this bot", cancellationToken: cancellationToken);
+                    break;
+            }
+        }
+
+        private Task<DialogTurnResult> ChoiceCardStep(DialogContext dc, WaterfallStepContext step, CancellationToken cancellationToken)
+        {
+            return dc.PromptAsync("cardPrompt", GenerateOptions(dc.Context.Activity), cancellationToken);
+        }
+
+        public static PromptOptions GenerateOptions(Activity activity)
+        {
+            return new PromptOptions()
```

Comment:
As a pattern, I think we should avoid this. 
var options = new PromptOptions() { ... }; 
...
return options. 

I find that kind of code much easier to read and to modify.